### DISCUSSION TITLE
realtime ClientResponseError 0

### DISCUSSION THREAD

**Discussion by babakfp** [pocketbase/pocketbase#1848] [Q&A] [answered]
Hey

Here is the error that I'm getting:
```
{"url":"","status":0,"data":{},"isAbort":false,"originalError":{},"name":"ClientResponseError 0"}
```
I'm only getting this error on my phone. On the desktop (dev and production), it works fine.

I'm using 2 instances of PocketBase. One on the server and another on the client side. I needed to create the second instance on the client because realtime doesn't work on server files because server files only run once (and run on page load/refresh). I'm using PocketBase with SvelteKit. Here is my realtime code in `+layout.svelte` file:

```html
<script>
	import { onMount } from "svelte"
	import PocketBase from "pocketbase"
	import { PUBLIC_POCKETBASE_URL } from "$env/static/public"
	import { messages } from "$lib/Chat/messages.js"

	const pb = new PocketBase(PUBLIC_POCKETBASE_URL)

	onMount(async () => {
		try {
			await pb
				.collection("messages")
				.subscribe("*", async ({ action, record }) => {
					if (action === "create") {
						const userRecord = await pb
							.collection("users")
							.getOne(record.user)
						record.expand = { user: structuredClone(userRecord) }
						messages.update(v => [record, ...v])
					} else if (action === "delete") {
						messages.update(v => v.filter(m => m.id !== record.id))
					}
				})
		} catch (_error) {
			console.log("Something went wrong in chat in real-time.")
			console.log(_error)
		}
	})
</script>
```

Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
Where the above code is executed - only in the browser, only on the server-side or both?

If the code is executed in a node env make sure that an `EventSource` polyfill is loaded as in https://github.com/pocketbase/js-sdk#nodejs-via-npm:

```js
// for server: npm install eventsource --save
import eventsource from 'eventsource';

global.EventSource = eventsource;
```

Since the error doesn't have a `url` prop, it is most likely caused by a client-side issue, but I'm not sure what exactly. `originalError` being an empty object is also strange. You can try to wrap the inner `await` call with `try/catch` to narrow down the trigger.

If you are not able to identify the issue, feel free to provide a minimal repo and reproducible steps and I'll try to debug it.

**Reply by babakfp** [reply]
It only runs in the browser (`onMount`). I removed extra stuff:

```HTML
<script>
	import { onMount } from "svelte"
	import PocketBase from "pocketbase"
	import { PUBLIC_POCKETBASE_URL } from "$env/static/public"
	import { messages } from "$lib/Chat/messages.js"

	const pb = new PocketBase(PUBLIC_POCKETBASE_URL)

	let errorObject
	let actionName

	onMount(async () => {
		try {
			await pb
				.collection("messages")
				.subscribe("*", async ({ action, record }) => {
					actionName = action
				})
		} catch (_error) {
			errorObject = _error
		}
	})
</script>

<div>errorObject:<br />{JSON.stringify(errorObject)}</div>
<br />
<div>actionName:<br />{actionName}</div>
```

Result in the browser:

![photo_2023-02-13_05-18-36](https://user-images.githubusercontent.com/44144724/218468855-fd4c4547-e37a-4731-8c4f-79b762601c88.jpg)

---

Using `EventSource` throws this error in the browser:

```
TypeError: util.inherits is not a function
    at node_modules/.pnpm/eventsource@2.0.2/node_modules/eventsource/lib/eventsource.js (eventsource.js:345:6)
    at __require (chunk-RSJERJUL.js?v=5de7f929:3:50)
    at eventsource.js:495:1
```

**Reply by ganigeorgiev** [reply]
The `EventSource` polyfill is needed only if you want to use the realtime service in a server-side env.

There is not enough context to help you debug it since I'm not able to reproduce it.

Are you sure that `PUBLIC_POCKETBASE_URL` is not empty?

Could you provide a minimal reproducible repo?

What is the mobile browser and version where the issue is reproducible?

**Reply by ganigeorgiev** [reply]
@babakfp The only way I'm able to reproduce the "empty" (_see below why it is empty_) `originalError` is if the SDK is not able to connect to the PocketBase server, aka. **`PUBLIC_POCKETBASE_URL` is incorrect or not accessible from the mobile device**.

To confirm that this is the case, you can try to `console.log(err.originalError)` (note that it is empty in your example because the original error is not `JSON.stringify`-able ~~for some reason~~ _see the update below_).

**Update:** JSON.stringify(err) prints the `originalError` as empty `{}` because [per the docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) only enumarable properties are visited printed:

> Only [enumerable own properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) are visited. **This means [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set), etc. will become "{}".** You can use the [replacer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_replacer_parameter) parameter to serialize them to something more useful. Properties are visited using the same algorithm as [Object.keys()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys), which has a well-defined order and is stable across implementations.

**Reply by babakfp** [reply]
I find something, when I stop `pocketbase.exe` and try to view the frontend, it throws the exact same error. I'm connected to the frontend by exposing my local project to the network that I'm connected to. And with my phone, I'm connected to the same network and I'm able to view that website/frontend. I was going to say, maybe this is the problem, that I'm not exposing `pocketbase.exe` server to the network, but then I realized I'm already able to see the other db data on the frontend and interact with it! So...

> Are you sure that PUBLIC_POCKETBASE_URL is not empty?

Yes.

Here is a fresh sveltekit repo:
https://github.com/babakfp/sveltekit-clientresponseerror-0

NodeJS Version: `18.14.0`
PNPM Version: `7.26.3`
NPM Version: `9.3.1`
Google Chrome Version on my phone: `109.0.5414.117`

1. Clone
2. `pnpm i`
3. Add `PUBLIC_POCKETBASE_URL` to `.env`
4. `pnpm dev --host`
5. Open the public network URL with your phone (I'm using a phone called `Shiomi lite 11`)
6. Now, whenever you open/refresh the page, you are going to see that error object.

This is the code from pocketbase "Export collections" page for the collection that I'm trying to use realtime on:

```
    {
        "id": "86i2iracs6xus87",
        "name": "messages",
        "type": "base",
        "system": false,
        "schema": [
            {
                "id": "i3icqadd",
                "name": "content",
                "type": "editor",
                "system": false,
                "required": true,
                "unique": false,
                "options": {}
            },
            {
                "id": "ezwcmpzp",
                "name": "user",
                "type": "relation",
                "system": false,
                "required": true,
                "unique": false,
                "options": {
                    "collectionId": "_pb_users_auth_",
                    "cascadeDelete": false,
                    "maxSelect": 1,
                    "displayFields": [
                        "username"
                    ]
                }
            }
        ],
        "listRule": "",
        "viewRule": "",
        "createRule": "user = @request.auth.id",
        "updateRule": null,
        "deleteRule": null,
        "options": {}
    }
```

**Reply by ganigeorgiev** [reply]
@babakfp Please try to `console.log(err.originalError)` to see what is the actual error.

Additionally, if `PUBLIC_POCKETBASE_URL` is `localhost` or `127.0.0.1` then it will not work on the phone because it targets the localhost of your device, not the host. You have to provide a public accessible address (eg. the ip of your local network).

I'm mostly sure that this is caused by the mobile device not being able to access the PocketBase server. To eliminate local network misconfiguration, you can also try using [ngrok](https://ngrok.com/) or other ssh tunneling tool to create a public address for your localhost.

**Reply by babakfp** [reply]
Yep, that was the problem, I used this to expose pocketbase to my network: https://github.com/pocketbase/pocketbase/discussions/435#discussioncomment-3643130

Thanks🧡
